runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -83,6 +83,62 @@ TR::TreeLowering::lowerValueTypeOperations(TR::Node* node, TR::TreeTop* tt)
       }
    }
 
+/**
+ * @brief Copy register dependencies between GlRegDeps node at exit points.
+ *
+ * This function is only intended to work with GlRegDeps nodes for exit points,
+ * (i.e. BBEnd, branch, or jump nodes) within the same extended basic block.
+ *
+ * Register dependencies are copied "logically", meaning that the actual node
+ * used to represent a dependency won't necessarily be copied. If the reg dep
+ * is represented by a PassThrough, then the node itself is copied and its
+ * child is commoned (so it's lifetime is exteneded; note that in correctly-formed
+ * IL, the child must also be the child of a reg store in the containing block).
+ * Otherwise, the dependency must be represented by a reg load, which must have
+ * come from the GlRegDeps node at the entry point and *must* be commoned
+ * (so it won't get copied).
+ *
+ * In addition, this function allows *one* register dependency to be changed
+ * (substituted). That is, if a register dependency is found under `sourceNode`
+ * for the same register that is set on `subsituteNode`, then `substituteNode`
+ * will be used instead of the dependency from `sourceNode`. Note that the
+ * reference of of `substituteNode` is incremented if/when it gets added. If
+ * `substituteNode` is NULL the no substitution will be attempted.
+ *
+ * @param targetNode is the GlRegDeps node that reg deps are copied to
+ * @param sourceNode is the GlRegDeps node that reg deps are copied from
+ * @param substituteNode is the reg dep node to substitute if a matching register is found in `sourceNode` (NULL if none)
+ */
+void
+copyExitRegDepsAndSubstitue(TR::Node* const targetNode, TR::Node* const sourceNode, TR::Node* const substituteNode)
+   {
+   for (int i = 0; i < sourceNode->getNumChildren(); ++i)
+      {
+      TR::Node* child = sourceNode->getChild(i);
+      if (substituteNode
+          && child->getLowGlobalRegisterNumber() == substituteNode->getLowGlobalRegisterNumber()
+          && child->getHighGlobalRegisterNumber() == substituteNode->getHighGlobalRegisterNumber())
```

Comment:
Under the hood, `getGlobalRegisterNumber()` is the same as `getLowGlobalRegisterNumber()`. For values that are too large to store in just one register (e.g. a double floating-point value on a 32-bit system), we use `getLowGlobalRegisterNumber()` and `getHighGlobalRegisterNumber()` to represent the pair of registers used to store the complete value (low bits and high bits, respectively). So, without knowing the type of value we're storing, we need to conservatively assume that two registers are used. If we happen to only need one register, this case is covered by the fact that `getLowGlobalRegisterNumber() == getGlobalRegisterNumber()`.

Reply:
> without knowing the type of value we're storing, we need to conservatively assume that two registers are used. 

If we need to cover the possibility of the node being stored in two registers, should the condition check be`"||"` ?
```c++
if (substituteNode
          && (child->getLowGlobalRegisterNumber() == substituteNode->getLowGlobalRegisterNumber() ||  // stored in one register 
              child->getHighGlobalRegisterNumber() == substituteNode->getHighGlobalRegisterNumber())) // stored in a second register 
```

Reply:
The condition needs to be `&&` because both the low and high registers must match if two registers are used (by convention and for consistency, I'm given to understand that the high register must also match if only register is used).